cluster/assemblestate/devices.go

```diff
@@ -0,0 +1,256 @@
+package assemblestate
+
+import (
+	"time"
+)
+
+// DeviceQueryTracker manages device identity information and query/response
+// lifecycle for unknown devices in the cluster.
+type DeviceQueryTracker struct {
+	// responses is used to send notifications that indicate that we are able to
+	// respond to a peer's query for device information.
+	responses chan struct{}
+
+	// queries is used to send notifications that indicate that we should query
+	// a peer for device information.
+	queries chan struct{}
+
+	// inflight keeps track of queries for device information that we have
+	// in-flight. We keep track of a mapping of device RDT to time at which the
+	// query was sent.
+	inflight map[DeviceToken]time.Time
+
+	// timeout is the amount of time we will wait before sending a query for
+	// device information again.
+	timeout time.Duration
+
+	// unknowns keeps track of which devices each peer has queried us for.
+	unknowns map[DeviceToken]map[DeviceToken]struct{}
+
+	// sources keeps track of which devices we know each peer knows about.
+	sources map[DeviceToken]map[DeviceToken]struct{}
+
+	// ids is our collection of device identities that we've heard from other
+	// peers.
+	ids map[DeviceToken]Identity
+}
+
+// NewDeviceQueryTracker creates a new DeviceQueryTracker with the given
+// identity, timeout for inflight queries, and initial data.
+func NewDeviceQueryTracker(self Identity, timeout time.Duration, data DeviceQueryTrackerData) DeviceQueryTracker {
+	dt := DeviceQueryTracker{
+		timeout:   timeout,
+		responses: make(chan struct{}, 1),
+		queries:   make(chan struct{}, 1),
+		unknowns:  make(map[DeviceToken]map[DeviceToken]struct{}),
+		sources:   make(map[DeviceToken]map[DeviceToken]struct{}),
+		inflight:  make(map[DeviceToken]time.Time),
+		ids: map[DeviceToken]Identity{
+			self.RDT: self,
+		},
+	}
+
+	// seed with any provided data
+	for _, identity := range data.IDs {
+		dt.Identify(identity)
+	}
+
+	for peer, devices := range data.Unknowns {
+		dt.Query(peer, devices)
+	}
+
+	for peer, devices := range data.Sources {
+		dt.UpdateSource(peer, devices)
+	}
+
+	return dt
+}
+
+// ResponsesAvailable returns a channel that signals when there are device query
+// responses ready to be sent to peers. After reading from this channel, it is
+// expected that all responses are handled via calls to
+// [DeviceQueryTracker.QueryResponses]. In the case that responses cannot be
+// handled, requeue them with [DeviceQueryTracker.RetryResponses].
+func (d *DeviceQueryTracker) ResponsesAvailable() <-chan struct{} {
+	return d.responses
+}
+
+// RetryResponses signals that device query responses should be retried. This
+// should be called when a response transmission fails.
+func (d *DeviceQueryTracker) RetryResponses() {
```

Comment:
should this be intergrated instead in the corresponding ack function returned below or a as 2nd closure? it's a bit strange that on success we call a function and on failure we need to call a method?